retrace/glretrace.py

```diff
@@ -196,6 +198,24 @@ def retraceFunctionBody(self, function):
 
         Retracer.retraceFunctionBody(self, function)
 
+        # If we have a query buffer or if we execute the query we have to loop until the
+        # query result is actually available to get the proper result - for the following
+        # execution if the query buffer is used or for the check to make sense, and if we
+        # just want to execute the query for timing purpouses we also should wait
+        # for the result.
+        if function.name.startswith('glGetQueryObject'):
+           print(r'    if (_query_buffer || retrace::queryHandling != retrace::QUERY_SKIP) {')
+           print(r'        auto query_result = call.arg(2).toArray();')
+           print(r'        assert(query_result && query_result->values.size() == 1);')
+           print(r'        auto expect = static_cast<decltype(retval)>(query_result->values[0]->toUInt());')
+           print(r'        if (call.arg(1).toUInt() == GL_QUERY_RESULT_AVAILABLE) {')
+           print(r'            if (expect == 1 && retval == 0)')
+           print(r'                goto wait_for_query_result;')
+           print(r'        } else if (!_query_buffer && expect != retval) {')
+           print(r'           std::cout << "Warning: query returned " << retval << " but trace contained " << expect << "\n";')
```

Comment:
let's use `retrace::warning` so this gets plugged into the UI correctly.

Reply:
Changed.